### PR TITLE
✨clusterctl: move shared objects

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR refactors clusterctl in order to support move when two or move clusters are sharing some objects (like e.g. Templates or LoadBalancers).

The PR takes a generic approach to solve this problem and the move sequence is defined by processing the ownerReference chain, so we ensure that a node is moved only after its owners.

Additionally, nodes are grouped, so each group of nodes can be moved in parallel. e.g.
- All the Clusters should be moved first because they have no OwenerReferences
- All the MachineDeployments should be moved second because they have OwenerReferences to clusters, which are moved as part of the first group
- then all the MachineSets should be moved second because they have OwenerReferences to MachineDeployments, which are moved as part of the second group
- etc.

**Which issue(s) this PR fixes**:
rif: #1729 

/area clusterctl
/assign @ncdc
/assign @vincepri

### CODE REVIEW

File: cmd/clusterctl/pkg/client/cluster/mover_test.go [hunk 1]
```diff
@@ -17,99 +17,389 @@ limitations under the License.
...
+				gotNodes := []string{}
+				for _, node := range gotGroup {
+					gotNodes = append(gotNodes, string(node.identity.UID))
+				}
+
+				sort.Strings(gotNodes)
+				sort.Strings(wantGroup)
+
+				if !reflect.DeepEqual(gotNodes, wantGroup) {
+					spew.Dump(gotNodes)
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover_test.go]:
```suggestion
```

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 2]
```diff
@@ -119,32 +119,113 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
...
+		// but only few of them are related to Clusters/Machines etc.
+		moveGroup := moveGroup{}
+		for _, n := range graph.getNodesWithClusterTenants() {
+			// If the node was already included in the moveSequence, skip it.
+			if moveSequence.hasNode(n) {
+				continue
+			}
+
+			// Check if all the ownerReferences are already included in the move sequence; if yes, add the node to move group,
+			// otherwise skit it (the node will be re-processed in the next group).
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover.go]:
```suggestion
			// otherwise skip it (the node will be re-processed in the next group).
```

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 3]
```diff
@@ -119,32 +119,113 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
...
+		s.nodesMap[n] = empty{}
+	}
+}
+
+func (s *moveSequence) hasNode(n *node) bool {
+	_, ok := s.nodesMap[n]
+	return ok
+}
+
+func (s *moveSequence) group(i int) moveGroup {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover.go]:
```suggestion
func (s *moveSequence) getGroup(i int) moveGroup {
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go [hunk 4]
```diff
@@ -45,35 +43,45 @@ type dependAttributes struct {
...
 	identity corev1.ObjectReference
 
-	// dependents contains the list of nodes that are owned by the current node.
-	// Nb. This is the reverse of metadata.OwnerReferences.
-	dependents map[*node]dependAttributes
+	// owners contains the list of nodes that are owned by the current node.
+	owners map[*node]ownerReferenceAttributes
 
-	// dependents contains the list of nodes that are soft-owned by the current node.
+	// softOwners contains the list of nodes that are soft-owned by the current node.
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/objectgraph.go]: might be worth explaining what is a soft-owner?

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go [hunk 5]
```diff
@@ -45,35 +43,45 @@ type dependAttributes struct {
...
 }
 
 // markObserved marks the fact that a node was observed as a concrete object.
 func (n *node) markObserved() {
 	n.virtual = false
 }
 
-func (n *node) addDependent(dependent *node, attributes dependAttributes) {
-	n.dependents[dependent] = attributes
+func (n *node) addDOwner(owner *node, attributes ownerReferenceAttributes) {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/objectgraph.go]:
```suggestion
func (n *node) addOwner(owner *node, attributes ownerReferenceAttributes) {
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go [hunk 6]
```diff
@@ -45,35 +43,45 @@ type dependAttributes struct {
 type node struct {
 	identity corev1.ObjectReference
 
-	// dependents contains the list of nodes that are owned by the current node.
-	// Nb. This is the reverse of metadata.OwnerReferences.
-	dependents map[*node]dependAttributes
+	// owners contains the list of nodes that are owned by the current node.
+	owners map[*node]ownerReferenceAttributes
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/objectgraph.go]: I wonder if we can merge the owners and softOwners maps by adding a field to `ownerReferenceAttributes` called `implicit` or something similar?

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 7]
```diff
@@ -119,32 +119,113 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
...
 		if err := setClusterPause(o.fromProxy, cluster, o.log); err != nil {
 			return err
 		}
+	}
+
+	// Define the move sequence by processing the ownerReference chain, so we ensure that a node is moved only after its owners.
+	// Additionally nodes are grouped, so bulk of nodes can be moved in parallel. e.g.
+	// - All the Clusters should be moved first (group 1, processed in parallel)
+	// - All the MachineDeployments should be moved second (group 1, processed in parallel)
+	// - then all the MachineSets, then all the Machines, etc.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/mover.go]: If you're going to do it in this order, I think we may need to retain cluster.spec.paused=true while everything is moving. After the move is done, we can clear paused. Otherwise, if you have e.g. a Cluster, <Infra>Cluster, MachineDeployment, KubeadmConfigTemplate, <Infra>MachineTemplate, the MachineDeployment controller in the target cluster potentially will start creating new MachineSets, which will in turn create new Machines instead of reusing the ones that have yet to move. Is this concern valid?

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go (hunk 6 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/objectgraph.go]: I would prefer to keep the two concepts separated because one mirrors what is in the Kubernetes API while the other is a Cluster API convention, but If you prefer I can merge them together.

File: cmd/clusterctl/pkg/client/cluster/mover.go (hunk 7 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: > The MachineDeployment controller in the target cluster potentially will start creating new MachineSets, which will, in turn, create new Machines instead of reusing the ones that have yet to move

@ncdc this was not the case in the former implementation of move/not observed in my simple tests.

However, if we prefer to retain cluster.spec.paused=true while everything is moving it should not be difficult to do and it puts us on the safe side.

The only downside I can think of is the fact that we will be forced to remove (or rethink) the "waitForMachine" check because it will never be satisfied before clearing paused, but if I remember well you already expressed some concerns on that...

Comment (ncdc) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: @detiber you have tons of pivot expertise - care to 👀?

Comment (detiber) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: +1 to retaining `cluster.spec.paused=true` through the duration of the pivoting process, otherwise we are likely to end up with race conditions between the creation of resources for things like MachineDeployments/MachineSets.

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go (hunk 6 above)
Comment (vincepri) [reply cmd/clusterctl/pkg/client/cluster/objectgraph.go]: It's ok, we can clean it up later

File: cmd/clusterctl/pkg/client/cluster/mover.go (hunk 7 above)
Comment (vincepri) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: +1 from me as well, seems like the safest approach

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: Ok, so I'm dropping MachineWait. Yay!

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 8]
```diff
@@ -119,32 +119,113 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
...
 		o.log.Info("Moving", "Cluster", cluster.identity.Name, "Namespace", cluster.identity.Namespace)
 
-		// Sets the pause field on the Cluster object in the source management cluster, so the controllers stop reconciling it.
 		if err := setClusterPause(o.fromProxy, cluster, o.log); err != nil {
 			return err
 		}
+	}
+
+	// Define the move sequence by processing the ownerReference chain, so we ensure that a node is moved only after its owners.
+	// Additionally nodes are grouped, so bulk of nodes can be moved in parallel. e.g.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: I will make a pass to make it clear in the comment where we are talking about nodes in the graphs, but I would like to keep the names as it is.

**Comment by fabriziopandini** [comment]
@ncdc @vincepri @detiber comments addressed, ready for another pass.

**Comment by fabriziopandini** [comment]
@ncdc @vincepri it seems that tests are flaking, probably due to parallel move. it will be ok for you removing the parallelism now? without "WaitForMachines", there is no real need for this additional level of complexity now...

**Comment by vincepri** [comment]
Go for it, we don't need it to be parallel now

**Comment by fabriziopandini** [comment]
@vincepri done

**Comment by vincepri** [comment]
@fabriziopandini 

Seems linter is unhappy

```
cmd/clusterctl/pkg/client/cluster/mover.go:244:32: Using the variable on range scope `nodeToCreate` in function literal (scopelint)
			return o.createTargetObject(nodeToCreate, toProxy)
			                            ^
cmd/clusterctl/pkg/client/cluster/mover.go:361:32: Using the variable on range scope `nodeToDelete` in function literal (scopelint)
			return o.deleteSourceObject(nodeToDelete)
```

**Comment by fabriziopandini** [comment]
@vincepri CI is happy again 😉

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 9]
```diff
@@ -839,7 +1078,65 @@ func TestObjectGraph_DiscoveryByNamespace(t *testing.T) {
...
+		name        string
+		fields      fields
+		wantSecrets map[string][]string
+	}{
+		{
+			name: "A cluster with a soft owned secret",
+			fields: fields{
+				objs: test.NewFakeCluster("ns1", "foo").Objs(),
+			},
+			wantSecrets: map[string][]string{ // wantSecrets is a map[Cluster.UID] --> list of UIDs
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]:
```suggestion
			wantSecrets: map[string][]string{ // wantSecrets is a map[node UID] --> list of soft owner UIDs
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 10]
```diff
@@ -839,7 +1078,65 @@ func TestObjectGraph_DiscoveryByNamespace(t *testing.T) {
...
+			}
+
+			graph.setSoftOwnership()
+
+			gotSecrets := graph.getSecrets()
+			if len(gotSecrets) != len(tt.wantSecrets) {
+				t.Fatalf("got = %d secrets, want %d", len(gotSecrets), len(tt.wantSecrets))
+			}
+
+			for _, cluster := range gotSecrets {
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]:
```suggestion
			for _, secret := range gotSecrets {
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 11]
```diff
@@ -839,7 +1078,65 @@ func TestObjectGraph_DiscoveryByNamespace(t *testing.T) {
...
+				wantObjects, ok := tt.wantSecrets[string(cluster.identity.UID)]
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]:
```suggestion
				wantObjects, ok := tt.wantSecrets[string(secret.identity.UID)]
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 12]
```diff
@@ -839,7 +1078,65 @@ func TestObjectGraph_DiscoveryByNamespace(t *testing.T) {
...
+				if !ok {
+					t.Fatalf("got = %s, not included in the expected secrect list", cluster.identity.UID)
+				}
+
+				gotObjects := []string{}
+				for softOwners := range cluster.softOwners {
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]:
```suggestion
				for softOwners := range secret.softOwners {
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 13]
```diff
@@ -839,7 +1078,65 @@ func TestObjectGraph_DiscoveryByNamespace(t *testing.T) {
...
+					gotObjects = append(gotObjects, string(softOwners.identity.UID))
+				}
+
+				sort.Strings(wantObjects)
+				sort.Strings(gotObjects)
+
+				if !reflect.DeepEqual(gotObjects, wantObjects) {
+					t.Fatalf("secret %s, got = %s, expected = %s", cluster.identity.UID, gotObjects, wantObjects)
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]:
```suggestion
					t.Fatalf("secret %s, got = %s, expected = %s", secret.identity.UID, gotObjects, wantObjects)
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 14]
```diff
@@ -849,47 +1146,166 @@ func Test_objectGraph_setSoftDependents(t *testing.T) {
...
+					"bootstrap.cluster.x-k8s.io/v1alpha3, Kind=DummyBootstrapConfig, ns1/cluster2-m1",
+					"/v1, Kind=Secret, ns1/cluster2-m1",
 				},
 			},
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			graph, err := getDetachedObjectGraphWihObjs(tt.fields.objs)
+			gb, err := getDetachedObjectGraphWihObjs(tt.fields.objs)
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]: Why this change?

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go [hunk 15]
```diff
@@ -849,47 +1146,166 @@ func Test_objectGraph_setSoftDependents(t *testing.T) {
...
 			if err != nil {
 				t.Fatal(err)
 			}
 
-			graph.setSoftDependents()
+			// we want ti check that soft dependent nodes are considered part of the cluster, so we make sure to call SetSoftDependants before SetClusterTenants
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]:
```suggestion
			// we want to check that soft dependent nodes are considered part of the cluster, so we make sure to call setSoftOwnership before setClusterTenants
```

File: cmd/clusterctl/pkg/client/cluster/objectgraph_test.go (hunk 14 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/objectgraph_test.go]: I'm getting a little bit obsessed by code readability 🤣 
If you prefer I can rollback

**Comment by fabriziopandini** [comment]
@ncdc comment addressed

**Comment by ncdc** [comment]
LGTM. @fabriziopandini please squash!

**Comment by fabriziopandini** [comment]
@ncdc done!
@vincepri for the final lgtm
